### PR TITLE
Codeflash RF-DETR instance segmentation optimizations

### PR DESCRIPTION
**Pull request by dkosowski87**
# Description
PR gathering the changes proposed by Codeflash team to the RF-DETR instance segmentation pipeline.

PRs included:
- Initial Pre/Post-processing Triton Kernel introduction and 2-Depth pipeline optimization - https://github.com/roboflow/inference/pull/2405
- Fallback on internal Triton JIT compilation errors - https://github.com/roboflow/inference/pull/2465
- Root coordinates fix on sparse RLE path - https://github.com/roboflow/inference/pull/2473
- Fix for FPS regression - https://github.com/roboflow/inference/pull/2475
- Fix for missing `PTXASError` in older Triton versions - https://github.com/roboflow/inference/pull/2481
- Missing attribute for `InstanceSegmentationPredictionDC` - https://github.com/roboflow/inference/pull/2484
- `Future`s in Batch object being resolved  - https://github.com/roboflow/inference/pull/2486
- Context identifier for frame metadata and predictions matching in 2-depth pipeline - https://github.com/roboflow/inference/pull/2488
- Timeouts for response resolution - https://github.com/roboflow/inference/pull/2489
- Clean up stream after non-default exit https://github.com/roboflow/inference/pull/2491
- Leave only `STRETCH_TO` mode in allowed resizes modes - https://github.com/roboflow/inference/pull/2493
- Restore deferred coordinate conversion - https://github.com/roboflow/inference/pull/2496
- Dockerfile changes - https://github.com/roboflow/inference/pull/2502
- Previous `CompactMask` path removed - https://github.com/roboflow/inference/pull/2509

# Issues and fixes
✅ Fixed ⚠️ In Progress 🚫 Not Fixed

| Description | Status |
|----------|-------|
| Triton JIT Compilation errors raised due to internals like `Missing c compiler` or `Missing ptxas` fail inference during runtime and are not handled gracefully | ✅ |
| Multi-step workflows fail as the `Future` object leaks outside the `RoboflowInstanceSegmentationModelBlockV3`. Downstream steps don't understand that object | ✅ |
| In a multi-step workflow, the last frame of the video, which in the `PipelinedWorkflowRunner` is flushed at the end, doesn't go through the downstream steps that are present after the `roboflow_core/roboflow_instance_segmentation_model@v3`. |  ✅  |
| Workflows that change the dimensionality before the `roboflow_core/roboflow_instance_segmentation_model@v3` fail due to that step not being able to handle a batch of images on the fast path | ✅  |
| In workflows where there is a step changing the coordinate system (`roboflow_core/relative_static_crop@v1`), before the `roboflow_core/roboflow_instance_segmentation_model@v3`, while downstream steps get the proper predictions, directly passing results from the `roboflow_core/roboflow_instance_segmentation_model@v3` seems to provide wrong coordinates | ✅ |
| After fixing multi-step workflow Future resolution a regression was introduced in FPS | ✅  |
| `PTXASError` is missing on Triton 3.2.0 so importing that error will fail | ✅  |
| `InstanceSegmentationPredictionDC` was missing `"mask_format": "polygon"` necessary for backwards compatibility | ✅  |
| `Future` object was not being resolved in output constructor when provided as an element of a `Batch` object | ✅  |
| Video frame metadata matched with predictions in depth-2 pipeline only by deque order. No `id` that would connect through the pipeline stages | ✅   |
| No timeout on `Future` object resolution | ✅   |
| Executor leak on abnormal teardown |  ✅  |
| Non stretch resize modes incorrectly advertised |  ✅  |

# Pre-deploy actions
- Registering TensorRT RF-DETR Instance Segmentation pre-trained model packages on production for Jetsons: **Not needed**: https://github.com/roboflow/roboflow-edge/pull/419 handles this automatically on devices.

# Supported Paths ❗ 

- The optimized path is included only when running workflows through the `InferencePipeline`.
- The improvements are only visible when using the **TensorRT** package of **RF-DETR Instance segmentation models**.
- Only packages with **static batch size** of `1` can currently run using the fast path
- Only `STRETCH_TO` resize mode for input pre-processing
- Only the `v3` version of the `roboflow_instance_segmentation_model` block

# Usage ❗ 

See [preliminary documentation](https://github.com/aseembits93/inference/blob/eb40b5e5a545ee59cc4fe4adbbf2e46a59a5c319/docs/workflows/pipeline-optimization.md) for detailed description.

`TLDR`: these are the environment variables need to be set to get the full speed up:
| ENV | Value |
|----------|-------|
|ENABLE_AUTO_CUDA_GRAPHS_FOR_TRT_BACKEND| true |
|INFERENCE_MODELS_RFDETR_TRITON_PREPROC_ENABLED | true |
|INFERENCE_MODELS_RFDETR_TRITON_POSTPROC_ENABLED | true |
|RFDETR_PIPELINE_DEPTH |2|

# On scaling the solution to other models

The current **depth-2** pipeline proves the performance win, but it is still too t**ightly coupled to the RF-DETR TRT inference path**. We should turn it into a generic stream-pipelining capability so other models and workflow blocks can use the same delayed-response/flush mechanism without copying RF-DETR-specific logic.
- Move stream-pipeline eligibility into the execution engine or compiled workflow metadata, so `workflows.py` only consumes a generic “can this workflow be stream-pipelined?” signal.
- Extract stream bookkeeping from `RoboflowInstanceSegmentationModelBlockV3` into a reusable controller or dependency, instead of embedding queues, flush handling, and async response logic in the block.
- Replace RF-DETR-specific config names with generic stream-pipeline config, plus an allowlist of model/model-runtime types that may use depth-2 scheduling.
- Generalize `_StreamPredictionContext` into a reusable context object that can carry block-specific finalization metadata without knowing about instance segmentation fields.
- Avoid tying stream pipelining to `inference_model_adapters`; define the async/stream contract closer to inference_models, so direct model usage can support it too.
- Split forward_async/future behavior into an explicit mixin or protocol rather than making it feel like a universal InstanceSegmentationModel capability.
- Move deque/generation bookkeeping into a dedicated scheduler object, making ordering, tail flush, and “one-frame delay” behavior easier to test independently.

This is a initial plan to be refined we discussed with @aseembits93 and @misrasaurabh1 that work on this will be started after merging this contribution so we don't block the speed-up introduction. Although this should be handled in order to scale to other models.

# Verification

#### Jetson Orin AGX, JP 6.2

| Property | Value |
|----------|-------|
| GPU | orin (CUDA CC 8.7) |
| NVIDIA driver | 540.4.0 |
| CUDA version | 12.6.68.post1 |
| TRT version | 10.7.0 |
| L4T version | 36.4.0 |
| torch | 2.6.0 |
| torchvision | 0.21.0 |
| ONNX runtime | 1.20.0 |
| triton | 3.2.0 |

#### Jetson Thor, JP 7.1

| Property | Value |
|----------|-------|
| GPU | nvidia-thor (CUDA CC 11.0) |
| NVIDIA driver | N/A |
| CUDA version | 13.0.48.post1 |
| TRT version | 10.13.3.9 |
| L4T version | 38.4.0 |
| torch | 2.10.0 |
| torchvision | 0.25.0 |
| ONNX runtime | 1.24.2 |
| triton | 3.6.0 |

#### T4

| Property | Value |
|----------|-------|
| GPU | Tesla T4 (CC 7.5) |
| NVIDIA driver | 550.163.1 |
| CUDA | 12.4.127.post1 |
| TensorRT | 10.12.0.36 |
| OS | ubuntu-22.04 |
| torch | 2.6.0 |
| ONNX Runtime | 1.21.1 |

### Single instance segmentation step workflow
`development/stream_interface/codeflash-optimization-compare.py`

Inference pipeline runs given a video `mp4` file (over 700 frames produced), through a workflow that only includes the `roboflow_core/roboflow_instance_segmentation_model@v3` block.

Workflow definition:
```json
  {
        "version": "1.0",
        "inputs": [{"type": "WorkflowImage", "name": "image"}],
        "steps": [
            {
                "type": "roboflow_core/roboflow_instance_segmentation_model@v3",
                "name": "segmentation",
                "images": "$inputs.image",
                "model_id": "rfdetr-seg-small",  // or rfdetr-seg-xlarge
                "confidence_mode": "custom",
                "custom_confidence": 0.4,
                "enforce_dense_masks_in_inference_models": false,
            },
        ],
        "outputs": [
            {
                "type": "JsonField",
                "name": "predictions",
                "selector": "$steps.segmentation.predictions",
            },
        ],
    }
```

#### rfdetr-seg-small on Jetson Orin AGX, JP 6.2

| Resolution | Baseline FPS | Optimized FPS | Speedup |
| :--------- | :----------- | :------------ | :------ |
| 640×360    | 20.43        | 67.04         | 3.28×   |
| 1920×1080  | 4.37         | 35.07         | 8.02×   |
| 3840×2160  | 1.26         | 18.74         | 14.93×  |

#### rfdetr-seg-xlarge on Jetson Thor, JP 7.1

| Resolution | Baseline FPS | Optimized FPS | Speedup |
| :--------- | :----------- | :------------ | :------ |
| 640×360    | 29.38        | 116.34        | 3.96×   |
| 1920×1080  | 7.40         | 51.74         | 6.99×   |
| 3840×2160  | 2.16         | 29.96         | 13.90×  |
---

### Multi-step workflow
`development/stream_interface/codeflash-optimization-compare-multistep-workflow.py`

Inference pipeline runs given a video `mp4` file (over 700 frames produced), through a workflow that includes an additional step before and after the `roboflow_core/roboflow_instance_segmentation_model@v3` block.

Workflow definition:
```json
  {
        "version": "1.0",
        "inputs": [{"type": "WorkflowImage", "name": "image"}],
        "steps": [
            {
                "type": "roboflow_core/relative_statoic_crop@v1",
                "name": "center_crop",
                "images": "$inputs.image",
                "x_center": 0.5,
                "y_center": 0.5,
                "width": 0.8,
                "height": 0.8,
            },
            {
                "type": "roboflow_core/roboflow_instance_segmentation_model@v3",
                "name": "segmentation",
                "images": "$steps.center_crop.crops",
                "model_id": "rfdetr-seg-small",  // or rfdetr-seg-xlarge,
                "confidence_mode": "custom",
                "custom_confidence": 0.4,
                "enforce_dense_masks_in_inference_models": false,
            },
            {
                "type": "roboflow_core/per_class_confidence_filter@v1",
                "name": "confidence_filter",
                "predictions": "$steps.segmentation.predictions",
                "class_thresholds": {},
                "default_threshold": 0.0,
            },
        ],
        "outputs": [
            {
                "type": "JsonField",
                "name": "predictions",
                "selector": "$steps.confidence_filter.predictions",
            },
            {
                "type": "JsonField",
                "name": "raw_predictions",
                "selector": "$steps.segmentation.predictions",
            },
        ],
    }
```

#### rfdetr-seg-small on Jetson Orin AGX, JP 6.2

| Resolution | Baseline FPS | Optimized FPS | Speedup |
| :--------- | :----------- | :------------ | :------ |
| 640×360    | 22.60        | 58.70         | 2.60x   |
| 1920×1080  | 5.85         | 21.26         | 3.64x   |
| 3840×2160  | 1.68         | 7.84          | 4.67x   |

#### rfdetr-seg-small on T4
(different multi-step workflow - instance segmentation -> mask visualisation -> vision events (+ parallel branch with clip embeddings)
| Profile   | Frames | Elapsed | FPS   |
|-----------|--------|---------|-------|
| Baseline  | 704    | 125.15s | 5.63  |
| Optimized | 704    | 32.37s  | 21.75 |
| **Speedup** | --   | --      | **3.87x** |

---

### Ablation study

Checking the effect of disabling one of the four type of optimizations introduced

#### rfdetr-seg-small on Jetson Orin AGX, JP 6.2

| Resolution | Baseline FPS | Optimized FPS | Speedup |
| :--------- | :----------- | :------------ | :------ |
| Depth-1   | 4.72        | 13.58         | 2.88x   |
| Triton post-process disabled | 4.39     | 9.02     | 2.06x |
| Triton pre-process disabled | 4.48 | 25.68 | 5.74x |
| CUDA Graph disabled | 4.56 | 13.47 | 2.96x |

### Warm-up
Triton JIT kernels generate GPU code during the first run, so the first frame inference is slower, around 12s.
latency-trace

### Parity check

`development/stream_interface/codeflash-optimization-compare-visualization-parity.py`

#### Numerical Comparison
Baseline and optimized profiles produced matching predictions within tolerance that was tested `1e-5`

##### Baseline
baseline_frame_000003

##### Optimized
optimized_frame_000003

### Memory check
On `t4`, using `rfdetr-seg-small` 1080p `vehicles` video.

| Metric | Main | Optimized | Delta |
|----------|----------|----------|----------|
| Peak live device memory | 1094.461 MB | 228.337 MB | -866.124 MB, -79.1% |
| Final live device memory | 979.370 MB | 113.246 MB | -866.124 MB, -88.4% |
| Device allocations | 1094.461 MB / 74 ops | 228.337 MB / 16 ops | much lower |
| Pinned allocations | 0 MB / 3 ops | 26.280 MB / 9 ops | higher |
| Host-to-device memcpy | 1014.053 MB / 3233 ops | 3408.846 MB / 547 ops | more volume, fewer ops |
| Device-to-device memcpy | 3912.796 MB / 6999 ops | 955.750 MB / 6543 ops | -75.6% volume |
| Device-to-host memcpy | 3.166 MB / 13148 ops | 225.343 MB / 4118 ops | more volume, fewer ops |

### CODE REVIEW

**Review by PawelPeczek-Roboflow — APPROVED** [review]

**Comment by PawelPeczek-Roboflow** [comment]
test failing due to bootstrapping issue with other change
